File: python/141-Linked-List-Cycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# region Floyd's Cycle Finding Algorithm (slow/fast pointers)


class Solution:
    # Space O(1)
    # Time O(n)
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False
        slow, fast = head, head.next

        while fast and fast.next:
            if slow == fast:
                return True
            slow = slow.next
            fast = fast.next.next
        return False


# endregion


# region Hash Table / Set Solution
class Solution:
    # Space O(n)
    # Time O(n)
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        mySet = set()
        while head is not None:
            if head in mySet:
                return True
            else:
                mySet.add(head)
                head = head.next
        return False


# endregion

File: c#/141-Linked-List-Cycle.cs
#region 2 Pointers
// Time 	O(n)
// Space 	O(1)
public class Solution
{
    public bool HasCycle(ListNode head)
    {
        if (head == null)
            return false;

        ListNode fast = head.next;
        ListNode slow = head;

        while (fast != null && fast.next != null)
        {
            if (fast == slow)
                return true;
            fast = fast.next.next;
            slow = slow.next;
        }

        return false;
    }
}
#endregion


#region Set
// Time 	O(n)
// Space 	O(n)
public class Solution
{
    public bool HasCycle(ListNode head)
    {
        if (head == null)
            return false;

        HashSet<ListNode> seen = new();

        while (head != null)
        {
            if (seen.Contains(head))
                return true;
            seen.Add(head);
            head = head.next;
        }

        return false;
    }
}
#endregion
// Definition for a singly-linked list.
public class ListNode
{
    public int val;
    public ListNode next;

    public ListNode(int x)
    {
        val = x;
        next = null;
    }
}
